25 root to arrow transform new

mweinberg2718 opened the pull request:
Work-in-progress PR for issue #25, to build a ROOT -> Arrow array converter for the C++ transformer and send it to a Kafka topic in RIVER. Basic workflow implemented, with correct readout to Kafka, but may need some improvements to efficiency.

mweinberg2718 committed:
Initial attempt to write Arrow arrays

mweinberg2718 committed:
Initial attempt to write Arrow arrays

mweinberg2718 committed:
Added Arrow output + Kafka broker

mweinberg2718 committed:
Added Arrow output + Kafka broker

mweinberg2718 committed:
Update to uproot classes

mweinberg2718 committed:
Change Kafka topic name to request ID

mweinberg2718 committed:
First commit

BenGalewsky commented on Dockerfile:
```diff
@@ -19,18 +19,20 @@ RUN sudo yum install -y xrootd-server xrootd-client xrootd \
+# RUN /bin/bash -c "source /home/atlas/release_setup.sh && \
+#     pip install --user --upgrade pip && \
+#     pip install --user -r requirements.txt && \
+#     pip install --user --index-url https://test.pypi.org/simple/ --no-deps servicex==0.2rc2"
```
Let's not leave commented out code in the repo - this can always be found in the commit history.

BenGalewsky commented on Dockerfile:
```diff
@@ -19,18 +19,20 @@ RUN sudo yum install -y xrootd-server xrootd-client xrootd \
-    validate_requests.py \
-    proxy-exporter.sh \
-    ./
+COPY . .
```
Wildcard copies can make the docker build's longer since they touch file timestamps and force Docker to rebuild earlier layers. If possible, just specify the files or directories you need

BenGalewsky commented on transformer.py:
```diff
@@ -84,7 +92,7 @@ def callback(channel, method, properties, body):
 
 
-def transform_single_file(file_path, output_path):
+def transform_single_file(file_path, output_path, servicex):
```
I don't think this will work in command line mode where the user provides a --path argument. I think this should default to `None` and then add some checks wherever you reference this to see if serviceX is defined

BenGalewsky commented on transformer.py:
```diff
@@ -114,10 +136,15 @@ def compile_code():
 
-    assert args.result_format == 'root-file', 'We only know how to create root file output'
-    assert args.result_destination != 'kafka', 'Kafka not yet supported'
+    # assert args.result_format == 'root-file', 'We only know how to create root file output'
```
Same note about commented out code

BenGalewsky commented on transformer.py:
```diff
@@ -98,6 +106,20 @@ def transform_single_file(file_path, output_path):
+    flat_tree_name = flat_file.keys()[0]
+    attr_name_list = flat_file[flat_tree_name].keys()
+
+    arrow_writer = ArrowWriter(file_format=args.result_format, servicex=servicex,
```
I think we need to preserve the ability to upload flat root files to object-store. This function will always turn the root file into arrow. Take a look at the old code to see how to upload the result directly to object-store

BenGalewsky reviewed (changes requested):
Excellent start - just some cleanup and making sure we can still run from the command line and keep the ability to upload root files to object-store

mweinberg2718 committed:
Removing commented-out code

mweinberg2718 committed:
Reverting to avoid wildcard COPY commands

mweinberg2718 committed:
servicex now defaults to None

Value of servicex now defaults to None; no additional checks are necessary since checking is performed in ArrowWriter class.

mweinberg2718 committed:
Removed commented-out code

mweinberg2718 committed:
Only send arrays to Kafka if no object store

mweinberg2718 committed:
Added retries (currently hardcoded to 3)

BenGalewsky commented:
This PR has been superseded by all of the work in the `102_uproot_transformer` branch. The commits have been picked up there.
